File: contests/codeforces-mar22-24/permutation.hs
-- input parsing

import Data.Char

stoi :: String -> [Int]
stoi "" = [0]
stoi str
  | isDigit x = init xs ++ [last xs * 10 + digitToInt x]
  | otherwise = xs ++ [0]
  where
    x = last str
    xs = stoi (init str)


-- math - factorial & binomial coef

mmul :: Int -> Int -> Int
mmul a b = a * b `mod` 1000000007

fact :: Int -> Int
fact 0 = 1
fact x = x `mmul` fact (x-1)

bexp :: Int -> Int -> Int
bexp a 0 = 1
bexp a p = bexp (a `mmul` a) (p `div` 2) `mmul` if even p then 1 else a

inv :: Int -> Int
inv x = bexp x 1000000005

nck :: Int -> Int -> Int
nck n k = num `mmul` inv den
  where
    num = fact n
    den = fact k `mmul` fact (n-k)


-- compute reductions of prefix/suffixes

reduce :: Int -> [Int] -> Int
reduce n s =
  if length s < 3 then 1
  else
    (reduce n (init s)) `mmul` (fact diff) `mmul` (nck (last s-2) diff)
    where
      diff = (last s) - (last (init s)) - 1
 
rev :: Int -> [Int] -> [Int]
rev n [] = []
rev n s = n+1 - last s : rev n (init s)


-- process test cases

test_case :: Int -> IO ()
test_case 0 = return ()
test_case x = do
  con <- getLine
  let [n,p,s] = stoi con
  prefix <- getLine
  let ps = stoi prefix
  suffix <- getLine
  let ss = stoi suffix
  if last ps /= head ss || 1 /= head ps || n /= last ss
    then print 0
    else
      print ((nck (n-1) (head ss - 1)) `mmul` (reduce n ps) `mmul`
        (reduce n (rev n ss)))
  test_case (x-1)


-- main

main :: IO ()
main = do
  cases <- getLine
  let c = (read cases :: Int)
  test_case c

File: contests/codeforces-mar22-24/permutation.py
M = 1000000007
P = 1000000005

def fact(x):
    ret = 1
    for i in range(2, x+1):
        ret = (ret*i) % M
    return ret

def inv(x):
    ret = 1
    p = P
    while p > 0:
        if p % 2 == 1:
            ret = (ret*x) % M
        x = (x*x) % M
        p = p // 2
    return ret

def nck(n, k):
    num = fact(n)
    den = (fact(k)*fact(n-k)) % M
    return (num*inv(den)) % M

def test_case():
    [n,p,s] = [int(x) for x in input().split()]
    ps = [int(x) for x in input().split()]
    ss = [int(x) for x in input().split()]

    if ps[-1] != ss[0] or ps[0] != 1 or ss[-1] != n:
        print(0)
        return

    ret = nck(n-1, ss[0]-1)

    for i in range(1, p):
        diff = ps[i] - ps[i-1] - 1
        if diff > 0:
            ret = (ret*nck(ps[i]-2, diff)) % M
            ret = (ret*fact(diff)) % M

    for i in range(1, s):
        diff = ss[i] - ss[i-1] - 1
        if diff > 0:
            ret = (ret*nck(n-ss[i-1]-1, diff)) % M
            ret = (ret*fact(diff)) % M

    print(ret)

t = int(input())

for i in range(t):
    test_case()
